articles/remapping.md

```diff
@@ -0,0 +1,77 @@
+
+ - Replace all of the basename with another basename
+ 
+- Arbitrary substring replacement
+
+ - Replace any substring with another string
+ 
+- Exact replacement (This is ROS 1)
+
+ - Replace a name with another name exactly
+
+ROS 2 may support remapping rules being applied either before or after the names are expanded to FQN. ROS 1 only supports the latter. 
+
+#### Pre-FQN expansion remapping
+Applying rules prior to FQN expansion enables this use case. Say the source code for a node use two names `"cat"` and `"/ns/cat"`. The node is in namespace `"ns"`. With just Post-FQN expantion it is impossible to remap these names individually. With Pre-FQN expansion remapping, a rule can be made to remap just one of them.
+
+#### Requirements for remapping capabilities
+##### 1. Namespace Replacement
+**A user can remap the first namespace token without replacing all namespace tokens with the same name.**
+
+*Example*
+Node uses names `"/big/cat"`, `"/big/red/dog"`, `"/red/big/cat"`. It must be possible to make a rule to that replaces "big" with "small" such that the final names are `"/small/cat"`, `"/small/red/dog"`, `"/red/big/cat"`
+
+**A user can replace all matching namespace tokens without changing the basename.**
+
+*Example*
+Node uses names `"/big/cat"`, `"/red/big/dog"`, `"/dog/big"`. It must be possible to make a rule to that replaces "big" with "small" such that the final names are `"/small/cat"`, `"/red/small/dog"`, `"/dog/big"`
+
+**A user can replace multiple matching namespace tokens only if whole namespaces are matched.**
+
+*Example*
+Node uses names `"/big/red/dog"`, `"/very_big/red/dog"`. It must be possible to make a rule to that replaces "big/red" with "small" such that the final names are `"/small/dog"`, `"/very_big/red/dog"`
+
+##### 2. Basename Replacement
+**A user can replace a full basename token with another basename token**
+
+*Example*
+Node uses names `"/big/cat"`, `"/big/cat/paws"`. It must be possible to make a rule to that replaces "cat" with "dog" such that the final names are `"/big/dog"`, `"/big/cat/paws"`
+
+##### 3. Exact Replacement
+**A user can replace a full name with another name**
+
+*Example*
+Node uses names `"/big/cat"`, `"/big/cat/paws"`. It must be possible to make a rule to that replaces "/big/cat" with "/bear" such that the final names are `"/bear"`, `"/big/cat/paws"`
+
+##### 4. Arbitrary Substring Replacement
+**A user can replace any substring with another string**
+
+*Examples*
+Node uses names `"/big/red/dog"`, `"/very_big/red/dog"`. It must be possible to make a rule to that replaces "big/red" with "small" such that the final names are `"/small/dog"`, `"/very_small/dog"`
+
+Node uses names `"/big/dog"`, `"/dog/big"`. It must be possible to make a rule to that replaces "dog" with "cat" such that the final names are `"/big/cat"`, `"/cat/big"`
+
+Node uses names `"/big/red/dog"`. It must be possible to make a rule to that replaces "/" with "_" such that the final name is `"/_big_red_dog"`
```

Comment:
I think that case might be exact name replacement. I think in ROS 1 `/big/red/dog:=_big_red_dog` would become `/ns/_big_red_dog`. I'll add multiple topics to this example to clarify how this remapping might operate.